account_invoice_pricelist/models/account_move.py

```diff
@@ -0,0 +1,111 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    pricelist_id = fields.Many2one(
+        comodel_name="product.pricelist",
+        string="Pricelist",
+        readonly=True,
+        states={"draft": [("readonly", False)]},
+    )
+
+    @api.constrains("pricelist_id", "currency_id")
+    def _check_currency(self):
+        for sel in self.filtered(lambda a: a.pricelist_id and a.is_invoice()):
+            if sel.pricelist_id.currency_id != sel.currency_id:
+                raise UserError(
+                    _("Pricelist and Invoice need to use the same currency.")
+                )
+
+    @api.onchange("partner_id", "company_id")
+    def _onchange_partner_id_account_invoice_pricelist(self):
+        if (
+            self.partner_id
+            and self.is_sale_document()
+            and self.partner_id.property_product_pricelist
+        ):
+            self.pricelist_id = self.partner_id.property_product_pricelist
+            self._set_pricelist_currency()
```

Comment:
No need to call "self._set_pricelist_currency()" here. When you set value for self.pricelist_id at line 32, "_set_pricelist_currency" will be called as well.